// cosmos-actor/lib/actorManager.js
'use strict';

var Promise = require('bluebird');
var logger = require('cosmos-logger').getLogger('cosmos');

var ActorManager = function ActorManager(actorConfig) {
	this.config = actorConfig;

	this.name = this.config.name;

	this.handler = new actorConfig.handler();
	this.remote = new actorConfig.remote();
};

ActorManager.prototype.init = function () {
	return new Promise(function (resolve, reject) {
		logger.info('=== ActorManager inited! ===');
		resolve();
	});
};

// be called by remote
ActorManager.prototype.onRpcCall = function (funcName, funcArgs) {
	var $this = this;
	return new Promise(function (resolve, reject) {
		var func = $this.remote.__proto__[funcName];
		if (!func) {
			reject(new Error('not found func: ' + funcName));
		} else {
			var ret = func.apply($this.remote, funcArgs);
			if (ret instanceof Promise) {
				return ret.then(function (resolveValue) {
					resolve(resolveValue);
				})['catch'](function (err) {
					reject(err);
				});
			} else {
				resolve(ret);
			}
		}
	});
};
module.exports = ActorManager;
